definitions/npm/react-relay_v1.x.x/flow_v0.47.x-/react-relay_v1.x.x.js

```diff
@@ -62,527 +112,167 @@ declare module "react-relay" {
-     */
-    check(selector: Selector): boolean;
-
-    /**
-     * Read the results of a selector from in-memory records in the store.
-     */
-    lookup(selector: Selector): Snapshot;
-
-    /**
-     * Notify subscribers (see `subscribe`) of any data that was published
-     * (`publish()`) since the last time `notify` was called.
-     */
-    notify(): void;
-
-    /**
-     * Publish new information (e.g. from the network) to the store, updating its
-     * internal record source. Subscribers are not immediately notified - this
-     * occurs when `notify()` is called.
-     */
-    publish(source: RecordSource): void;
-
-    /**
-     * Attempts to load all the records necessary to fulfill the selector into the
-     * target record source.
-     */
-    resolve(
-      target: MutableRecordSource,
-      selector: Selector,
-      callback: AsyncLoadCallback
-    ): void;
-
-    /**
-     * Ensure that all the records necessary to fulfill the given selector are
-     * retained in-memory. The records will not be eligible for garbage collection
-     * until the returned reference is disposed.
-     */
-    retain(selector: Selector): Disposable;
-
-    /**
-     * Subscribe to changes to the results of a selector. The callback is called
-     * when `notify()` is called *and* records have been published that affect the
-     * selector results relative to the last `notify()`.
-     */
-    subscribe(
-      snapshot: Snapshot,
-      callback: (snapshot: Snapshot) => void
-    ): Disposable;
-  }
+  declare export type $FragmentRef<T> = {
+    +$fragmentRefs: $PropertyType<T, '$refType'>,
+  };
 
   /**
-   * An interface for imperatively getting/setting properties of a `Record`. This interface
-   * is designed to allow the appearance of direct Record manipulation while
-   * allowing different implementations that may e.g. create a changeset of
-   * the modifications.
+   * A utility type that takes the Props of a component and the type of
+   * `props.relay` and returns the props of the container.
    */
-  declare export interface RecordProxy {
-    copyFieldsFrom(source: RecordProxy): void;
-
-    getDataID(): DataID;
-
-    getLinkedRecord(name: string, args?: ?Variables): RecordProxy;
-
-    getLinkedRecords(name: string, args?: ?Variables): ?Array<?RecordProxy>;
-
-    getOrCreateLinkedRecord(
-      name: string,
-      typeName: string,
-      args?: ?Variables
-    ): RecordProxy;
-
-    getType(): string;
-
-    getValue(name: string, args?: ?Variables): mixed;
-
-    setLinkedRecord(
-      record: RecordProxy,
-      name: string,
-      args?: ?Variables
-    ): RecordProxy;
-
-    setLinkedRecords(
-      records: Array<?RecordProxy>,
-      name: string,
-      args?: ?Variables
-    ): RecordProxy;
+  declare export type $RelayProps<Props, RelayPropT> = $ObjMap<
+    $Diff<Props, { relay: RelayPropT | void }>,
+    (<T: { +$refType: empty }>(T) => T) &
+      (<T: { +$refType: empty }>(?T) => ?T) &
```

Comment:
i can only get `createFragmentContainer` to pass flow if i comment out these two `$refType: empty` lines. If i leave them in I get all kinds of errors about missing properties. It seems `createFragmentContainer` is returning the real prop types used by the contained component, rather than converting the types to fragmentRefs. Everything works great when those lines are removed however. Something to do with the `empty` type perhaps? I don't fully understand how `empty` works, but it kind of looks like `{ +$refType: empty }` will match anything, meaning those `{ +$refType: empty }` lines will cause any prop that is a fragment to be fully typed on `createFragmentContainer` rather than hitting one of the appropriate `<TRef: FragmentReference, T: { +$refType: TRef }>` lines and being converted to a dumb fragmentRef.

Reply:
Opened an issue about my theory here: https://github.com/facebook/relay/issues/2443